(Huge) Improved logic between `Application Cache` <-> `ChatHubService` <-> `MessageService` <-> `SignalRService` <-> `ChatHub`

JayKim94 opened the pull request:
# Before this PR can be merged, the following checkboxes have to be ticked

- [x] This code belongs to the backend and is tested, or tests are not useful
- [x] New functions with more than 5 lines have docstrings
- [x] New code has been integrated with the `MessengerService` if appropiate
- [x] New code has been integrated with the `SignalRService` if appropiate
- [x] Changes introduce no commented out code or `Console.Writeline()` statements used for debugging

# Frontend Changes

- App has now internal static variables which can be accessed globally
- The properties are `EventProvider` and `StateProvider`
- `StateProvider` provides application cache/states which can be written/retrieved via `CacheQuery`
- `StateProvider` loads initial data of `Team`, `PrivateChat` and `Message`, which was previously done by `ChatHubService`
- `ChatHubService` was completely removed and `MessengerService` is accessed directly by `TeamManager`, `MessageManager` and ui commands
- `TeamManager` and `MessageManager` handle cache data for `Team` and `Message` respectively and dwell in `StateProvider`
- `EventProvider` listens to `SignalRService` and manages related UI Event Handling
- `StateProvider.CurrentUser` will be assigned by `UserDataService` upon log-in and can be accessed throughout the app

- Implemented UI `ManageTeamRolesDialog`

# Backend Changes

- Renamed methods in `MessengerService`, `SignalRService` and `ChatHub` to be consistent throughout the application
- `Message` has now `TeamId` property (DB, Service, Tests all updated)
- Operations on `Role` and `Permission` in `TeamService` returns the id of the row


### Everything else works just like before, this needed to be done sooner or later
![image](https://user-images.githubusercontent.com/68469414/126678295-e12cc661-adc3-4abc-9e09-cea7f6cd65ed.png)
![image](https://user-images.githubusercontent.com/68469414/126719092-e96a564d-1a68-4c21-b8be-44844c28f479.png)

JonasMuehlmann commented:
@JayKim94 Why do messages have teamIds now?

JayKim94 commented:
> @JayKim94 Why do messages have teamIds now?

The problem was that user connections in `Signal-R-Hub` are registered with `TeamId` (it must be btw, because of other logics) but `RecipientId` of a `Message` directs to `ChannelId`. It complicates too much without gain if we register each user with each channel id.

JonasMuehlmann commented on Messenger/Messenger.Core/Services/TeamService.cs:
```diff
@@ -433,8 +433,8 @@ public static async Task<IList<Channel>> GetAllChannelsByTeamId(uint teamId)
-        /// <returns>True if successful, false otherwise</returns>
-        public static async Task<bool> AddRole(string role, uint teamId)
+        /// <returns>Id of the added team role</returns>
+        public static async Task<uint> AddRole(string role, uint teamId)
```
Make return type nullable to be consistent with the rest of the codebase.

JayKim94 replied:
resolved in `e58447f`

JonasMuehlmann commented on Messenger/Messenger.Core/Services/TeamService.cs:
```diff
@@ -458,7 +459,7 @@ INSERT INTO
         /// <param name="teamId">The id of the team to remove the role from</param>
         /// <returns>True if successful, false otherwise</returns>
-        public static async Task<bool> RemoveRole(string role, uint teamId)
+        public static async Task<uint> RemoveRole(string role, uint teamId)
```
Make return type nullable to be consistent with the rest of the codebase.

JayKim94 replied:
resolved in `e58447f`

JonasMuehlmann commented:
> > @JayKim94 Why do messages have teamIds now?
> 
> The problem was that user connections in `Signal-R-Hub` are registered with `TeamId` (it must be btw, because of other logics) but `RecipientId` of a `Message` directs to `ChannelId`. It complicates too much without gain if we register each user with each channel id.

And implementing a function like `GetTeamIdFromChannelId()` would not be cleaner?

JayKim94 commented:
> > > @JayKim94 Why do messages have teamIds now?
> > 
> > 
> > The problem was that user connections in `Signal-R-Hub` are registered with `TeamId` (it must be btw, because of other logics) but `RecipientId` of a `Message` directs to `ChannelId`. It complicates too much without gain if we register each user with each channel id.
> 
> And implementing a function like `GetTeamIdFromChannelId()` would not be cleaner?
> 
> Edit: Could you also take a look at the broken tests?

I reverted all the changes and implemented a different way to broadcast from the signal hub. I'm not sure what's wrong about the tests

JonasMuehlmann commented:
> > > > @JayKim94 Why do messages have teamIds now?
> > > 
> > > 
> > > The problem was that user connections in `Signal-R-Hub` are registered with `TeamId` (it must be btw, because of other logics) but `RecipientId` of a `Message` directs to `ChannelId`. It complicates too much without gain if we register each user with each channel id.
> > 
> > 
> > And implementing a function like `GetTeamIdFromChannelId()` would not be cleaner?
> > Edit: Could you also take a look at the broken tests?
> 
> I reverted all the changes and implemented a different way to broadcast from the signal hub. I'm not sure what's wrong about the tests

The broken tests are due to a DB change I made for my work in #210. I guess, we can ignore them, since they are already fixed in my branch, AFAIK.

JayKim94 commented:
> > > > > @JayKim94 Why do messages have teamIds now?
> > > > 
> > > > 
> > > > The problem was that user connections in `Signal-R-Hub` are registered with `TeamId` (it must be btw, because of other logics) but `RecipientId` of a `Message` directs to `ChannelId`. It complicates too much without gain if we register each user with each channel id.
> > > 
> > > 
> > > And implementing a function like `GetTeamIdFromChannelId()` would not be cleaner?
> > > Edit: Could you also take a look at the broken tests?
> > 
> > 
> > I reverted all the changes and implemented a different way to broadcast from the signal hub. I'm not sure what's wrong about the tests
> 
> The broken tests are due to a DB change I made for my work in #210. I guess, we can ignore them, since they are already fixed in my branch, AFAIK.

can we then merge?

JonasMuehlmann commented:
> > > > > > @JayKim94 Why do messages have teamIds now?
> > > > > 
> > > > > 
> > > > > The problem was that user connections in `Signal-R-Hub` are registered with `TeamId` (it must be btw, because of other logics) but `RecipientId` of a `Message` directs to `ChannelId`. It complicates too much without gain if we register each user with each channel id.
> > > > 
> > > > 
> > > > And implementing a function like `GetTeamIdFromChannelId()` would not be cleaner?
> > > > Edit: Could you also take a look at the broken tests?
> > > 
> > > 
> > > I reverted all the changes and implemented a different way to broadcast from the signal hub. I'm not sure what's wrong about the tests
> > 
> > 
> > The broken tests are due to a DB change I made for my work in #210. I guess, we can ignore them, since they are already fixed in my branch, AFAIK.
> 
> can we then merge?

I think, I can get #210 done today.

JayKim94 commented:
Additional changes include clean-ups, more comments, and the removal of `ChatHubService` which became obsolete after implementing `EventProvider` and `StateProvider`
